qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,469 @@
+        else:
+            qrom_control, soqs["rot_reg"] = bb.add(
+                qrom, control=qrom_control, target0_=soqs["rot_reg"]
+            )
+        return qrom_control, soqs
+
+
+# @attrs.frozen
+class RotationTree:
+    r"""Used by ControlledStatePreparationUsingRotations to get the corresponding rotation
+    angles.
+
+    The rotation angles are used to encode the amplitude of a state using the method described in
+    [1], section on arbitrary quantum state preparation, page 3.
+
+    The only methods to be used externally are extract_ROM_values_from_state, angle_2_ROM_value and
+    build_rotation_tree_from_state.
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    @staticmethod
+    def extract_ROM_values_from_state(state: ArrayLike, rot_reg_size: int, uncompute: bool = False):
+        r"""Gives a list of the ROM values to be loaded for preparing the amplitudes of a state.
+
+        The ith element of the returned list is another list with the rom values to be loaded when
+        preparing the amplitudes of the ith qubit for the given state.
+        """
+        rotation_tree = RotationTree.build_rotation_tree_from_state(state)
+        next_layer = [rotation_tree]
+        rom_vals = []
+        while len(next_layer) != 0:
+            this_layer = next_layer
+            next_layer = []
+            rom_vals_this_layer = []
+            for tree in this_layer:
+                angle = tree.angle_0
+                if uncompute:
+                    angle = 2 * np.pi - angle
+                rom_val = RotationTree.angle_2_ROM_value(angle, rot_reg_size)
+                rom_vals_this_layer.append(rom_val)
+                if tree.branch0 is not None:
+                    next_layer.append(tree.branch0)
+                if tree.branch1 is not None:
+                    next_layer.append(tree.branch1)
+            rom_vals.append(rom_vals_this_layer)
+        return rom_vals
+
+    @staticmethod
+    def build_rotation_tree_from_state(state):
+        r"""Given a list of coefficients, returns a tree-like object that contains the angles for
+        the rotations when preparing the state.
+        """
+        if len(state) == 2:
+            return RotationTree(abs(state[0]) ** 2, abs(state[1]) ** 2, None, None)
+        dn_l = RotationTree.build_rotation_tree_from_state(state[: len(state) // 2])
+        dn_r = RotationTree.build_rotation_tree_from_state(state[len(state) // 2 :])
+        return RotationTree(dn_l.sum_total, dn_r.sum_total, dn_l, dn_r)
+
+    @property
+    def angle_0(self):
+        r"""Angle that corresponds to p_0."""
+        return 2 * np.arccos(np.sqrt(self._p0))
+
+    @staticmethod
+    def angle_2_ROM_value(angle, rot_reg_size):
+        r"""Returns the value to be loaded to a QROM to encode the given angle with a certain value
```

Comment:
nit: single line summaries.

Reply:
also sorry for the nits but I think we aim for all lowercase variable names / function names. I would also replace _2_ with _to_.